Switch to typescript

abumalick opened the pull request:
Would you accept a PR to switch to typescript?

I can make it more readable inchallah. I was a bit aggressive on this one:
- I formatted most files using Prettier
- I switched to pika to build the package
- I removed the build file from the repository

I can remove some of these and make it more readable (commit after renaming js files to ts so that you can check actual changes).

I wanted to ask you before working on this actually.

abumalick commented:
@z3bi السلام عليكم
What do you think about this?

z3bi commented:
@abumalick this is awesome! I'm 100% onboard with turning this into typescript. 

I'm not familiar with pika, what are the advantages of that compared to other package managers?

I think it should be ok to remove the built file from the repo, but we would want to update the ci pipeline to generate an artifact with releases. We will probably also want to keep the import the same if possible to avoid breaking changes.

abumalick commented:
Happy that you are onboard!

You can read on pika here for example: https://www.pika.dev/blog/introducing-pika-pack
Basically it simplifies the configuration for creating packages. Also @FredKSchott that is maintaining is working a lot on having packages supporting new browser features, so they will keep optimizing it without needing to us to dig into the building process. We can just use it and not think too much about it. It is using babel and rollup but is simplifying the configuration for packages maintainers.

It is noted for the built file, I will check that in the end.

I will hard push more organized commits to show you the actual changes to the files. I am trying to not have any breaking changes.

abumalick commented on .github/workflows/node.js.yml:
```diff
@@ -83,14 +82,13 @@ jobs:
     steps:
       - uses: actions/checkout@v2
-        with:
-          token: ${{ secrets.GIT_ACCESS_TOKEN }}
```
I believe this was not used, I don't really understand why it is there actually

abumalick commented on .github/workflows/node.js.yml:
```diff
@@ -17,20 +17,17 @@ jobs:
 
     strategy:
       matrix:
-        node-version: [10.x, 12.x, 14.x, 15.x]
```
I am dropping version 10 and 15 which is not supported anymore https://nodejs.org/en/about/releases/

abumalick replied:
it allows us to upgrade to npm 8

abumalick commented:
@z3bi You can check https://github.com/batoulapps/adhan-js/pull/93/commits/788bd2d35f7d941087f31d8cb47650e9467734a7 to see the exact changes to the files that I pushed to make it work with typescript.

I dropped pika pack in the end, in favor of https://github.com/developit/microbundle (pika didn't have a new release for 2 years). It does basically the same, it handles most of the configuration by itself.

> I think it should be ok to remove the built file from the repo, but we would want to update the ci pipeline to generate an artifact with releases. We will probably also want to keep the import the same if possible to avoid breaking changes.

I added the build to the release job. It should work, I didn't test it yet through, the pipeline passes in my github but I didn't increase the version so no release for now. https://github.com/abumalick/adhan-js/actions/runs/2089942331

So the only change to the final package is that the built files are inside the dist folder, we have now multiple build files (commonJs, umd, etc...)

```json
"exports": {
    "require": "./dist/Adhan.cjs",
    "default": "./dist/Adhan.modern.js"
  },
  "main": "./dist/Adhan.cjs",
  "umd:main": "dist/Adhan.umd.js",
  "module": "./dist/Adhan.module.js",
  "unpkg": "./dist/Adhan.umd.js",
  "types": "./dist/Adhan.d.ts",
```

I believe it should not break anyone's usage of the package, but I would at least bump a minor version as it has important changes and might break in some extreme cases. What do you think? Should I bump to 4.4.0 or 5.0.0?

I will try to test the release assets when you tell me what is the version that I should use inchallah

abumalick commented:
@z3bi Do you have time to check my comments please ?

sgtsquiggs commented:
Hi @abumalick, I've already got a typescript rewrite in progress. When it is ready I'll post it and we can compare/contrast.

abumalick commented:
This PR is not a rewrite, @sgtsquiggs , it is more a transition to typescript. I only added the types that are necessary and didn't change the code structure.
There is some refactoring that could be done for a prettier typescript version, like use enums instead of objects.

You can check the typescript adaptation in this commit: https://github.com/batoulapps/adhan-js/pull/93/commits/6f17cae1ddabdbbd5e7b1fbc1f892ed12ab3357c

abumalick commented:
@z3bi Any follow up?

sgtsquiggs commented:
This would be easier to review if the js -> ts conversion was registered as a diff instead of a delete/create - possibly the eslint/prettier rules caused too many changes for git to automatically associate the two files as related. Could you submit prettier/eslint change as a separate MR that we could merge before this ts MR?

sgtsquiggs commented:
Also can you explain the need to rename the config file extensions `js` -> `cjs` ? This seems uncommon to me.

sgtsquiggs commented:
For packaging - instead of webpack/pika/microbundle I would prefer to use rollup. I was in the process of updating our bundling:

`rollup.config.js`
```js
import babel from "@rollup/plugin-babel";
import resolve from "@rollup/plugin-node-resolve";
import { terser } from "rollup-plugin-terser";

const extensions = [".js", ".ts"];

export default {
  input: "src/Adhan.js",
  output: [
    {
      file: "lib/bundles/bundle.esm.js",
      format: "esm",
      sourcemap: true,
    },
    {
      file: "lib/bundles/bundle.esm.min.js",
      format: "esm",
      plugins: [terser()],
      sourcemap: true,
    },
    {
      file: "lib/bundles/bundle.umd.js",
      format: "umd",
      name: "myLibrary",
      sourcemap: true,
    },
    {
      file: "lib/bundles/bundle.umd.min.js",
      format: "umd",
      name: "myLibrary",
      plugins: [terser()],
      sourcemap: true,
    },
  ],
  plugins: [
    resolve({ extensions }),
    babel({
      babelHelpers: "bundled",
      exclude: "./node_modules/**",
      extensions,
      include: ["src/**/*.js"],
    }),
  ],
};
```

With these `package.json` scripts:
```js
    "clean": "rimraf lib",
    "build:esm": "cross-env BABEL_ENV=esmUnbundled babel src --extensions '.js' --out-dir 'lib/esm' --source-maps",
    "build:cjs": "cross-env BABEL_ENV=cjs babel src --extensions '.js' --out-dir 'lib/cjs' --source-maps",
    "build:bundles": "cross-env BABEL_ENV=esmBundled rollup -c",
    "build": "npm-run-all -l clean -p build:esm build:cjs build:bundles",
```

abumalick commented:
@sgtsquiggs @z3bi Can I drop the Math.trunc polyfill that is in Astronomical file?
![image](https://user-images.githubusercontent.com/8377185/164035560-6adc3707-ce97-4a59-945f-4dfc875d0fa5.png)

abumalick commented:
I believe this PR is ready for review @z3bi @sgtsquiggs 🚀

sgtsquiggs commented:
Can you `npm remove -D @types/estree @types/jest` I believe these are unnecessary

sgtsquiggs commented:
> @sgtsquiggs @z3bi Can I drop the Math.trunc polyfill that is in Astronomical file?
> ![image](https://user-images.githubusercontent.com/8377185/164035560-6adc3707-ce97-4a59-945f-4dfc875d0fa5.png)

100%

sgtsquiggs commented on jest.config.js:
```diff
@@ -1,7 +1,7 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
is this necessary?

abumalick replied:
it is not required, it is from official documentation and give you some completion.
<img width="630" alt="image" src="https://user-images.githubusercontent.com/8377185/164111140-1ce16285-b466-452f-b0a3-cb7499c6700f.png">

Do you want to remove it?

sgtsquiggs replied:
Oh nice 👍

sgtsquiggs reviewed (commented):
.eslint.js can be replaced with:
```javascript
module.exports = {
  parser: '@typescript-eslint/parser',

  parserOptions: {
    ecmaVersion: 2018,
    project: ['./tsconfig.json'],
    sourceType: 'module',
  },

  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:jest/recommended',
    'plugin:jest/style',
    'plugin:prettier/recommended',
  ],

  overrides: [
    {
      files: ['*.js'],
      parser: null,
      parserOptions: { project: null },
      rules: {
        '@typescript-eslint/ban-ts-comment': 'off',
        '@typescript-eslint/no-unnecessary-condition': 'off',
      },
    },
  ],

  env: {
    node: true,
    browser: true,
    'jest/globals': true,
  },

  rules: {
    '@typescript-eslint/ban-ts-comment': 'warn',
    '@typescript-eslint/no-unnecessary-condition': 'warn',
    complexity: ['warn', 10],
    'max-lines': ['warn', 300],
    'max-params': ['warn', 5],
    eqeqeq: ['error', 'smart'],
    'no-var': 'error',
    'prefer-const': 'error',
    'object-shorthand': 'error',
  },

  plugins: ['@typescript-eslint', 'jest'],
};
```

and then you can drop the second tsconfig

sgtsquiggs reviewed (changes requested):
Last thing - can you  `npm remove -D @types/estree @types/jest` I believe these are unnecessary

sgtsquiggs commented:
> I believe `@types/jest` is necessary to give you typescript support for jest utilities in test files.

👍 

> For `@types/estree`, I am getting an error when it is not there, I can try again inchallah but it is the only way I found to fix that error.

I had ran into the same estree error as you - I was able to get it to stop appearing without adding `@types/estree` but I forgot what exactly I did. Perhaps we just merge this and fix later.

abumalick commented:
Thank you very much @sgtsquiggs 👍🏻
